compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.td

```diff
@@ -1768,4 +1768,43 @@ def FLOW_CollectiveReduceScatterOp : FLOW_Op<"collective.reduce_scatter", [
   ];
 }
 
+def FLOW_CollectivePermuteOp : FLOW_Op<"collective.permute", [
+  AllTypesMatch<["source", "target", "result"]>,
+  DeclareOpInterfaceMethods<Util_TiedOpInterface, [
+    "getTiedResult",
+    "getTiedResultOperandIndex",
+    "getTiedResultOperandIndices",
+  ]>,
+]> {
+  let summary = [{performs collective-permute operation}];
+  let description = [{The operation sends data to one rank and receives data from another.}];
+
+  let arguments = (ins
+    FLOW_CollectiveElementTypeAttr:$element_type,
+    FLOW_Tensor:$target,
+    FLOW_ShapeDynamicDims:$target_dims,
+    FLOW_Tensor:$source,
+    FLOW_Channel:$channel,
+    I32:$param,
```

Comment:
we could instead take the send/recv ranks here instead of folding into the param and let the param folding happen in flow->stream - that keeps this level more readable and easier to lower into from other dialects and leaves the packing as an implementation detail in the compiler

Reply:
+1. We can regard the flow level as the user interfacing level, so readability is important.